게시물 좋아요 API 작성

Chestnut90 opened the pull request:
## PR 체크리스트
아래 항목을 확인해 주세요:

- [x] 커밋 메시지가 우리의 가이드라인을 따르고 있는지 확인하세요
- [x] 변경 사항에 대한 테스트가 추가되었는지 확인하세요 (버그 수정 / 기능 추가)
- [ ] 문서가 추가되거나 업데이트되었는지 확인하세요 (버그 수정 / 기능 추가)

## PR 유형
이 PR은 어떤 종류의 변경을 가져오나요?

<!-- 이 PR에 해당하는 것을 "x"로 표시하세요. -->

- [ ] 버그 수정
- [x] 새로운 기능 추가
- [ ] 코드 스타일 업데이트 (서식, 로컬 변수)
- [ ] 리팩터링 (기능 변경 없음, API 변경 없음)
- [ ] 빌드 관련 변경
- [ ] CI 관련 변경
- [ ] 문서 내용 변경
- [ ] 애플리케이션 / 인프라 변경
- [ ] 기타... 설명:

## 현재 동작은 무엇인가요?
<!-- 수정 중인 현재 동작을 설명하거나 관련된 이슈에 링크하세요. -->

이슈 번호: #11 

## 새로운 동작은 무엇인가요?
- 게시물(post)에 대한 좋아요(likes) api 추가 및 테스트

## 이 PR은 호환성 변경을 도입하나요?

- [ ] 예
- [x] 아니요

<!-- 이 PR에 호환성 변경이 포함되어 있다면, 기존 응용 프로그램에 대한 영향과 마이그레이션 경로를 아래에 설명해 주세요. -->

## 기타 정보

Chestnut90 committed:
:sparkles:Feat : 게시물 좋아요 api #11

- 게시물 좋아요, /likes/<content_id>/ 의 post method에 대한 APIView
Related to #11

Chestnut90 committed:
:white_check_mark:Test : 게시물 좋아요 api 테스트 #11

- authenticated 되지 않은 경우Unauthorized(401)
- authenticated 된 경우 like_count를 증가하며 Ok(200)

Related to #11

Chestnut90 committed:
:fire:Remove : 게시물 좋아요 api 관련 불필요 코드 제거 #11

Related to #11

JaeHyuckSa commented:
코드 작성하시느라 고생하셨습니다 ! :)

lfoyh6591 commented:
post_type에 따른 url 분리가 구현이 안된 것 같습니다.
![image](https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/assets/73338772/c043e5f8-ad34-4624-b8ae-e8f09dcaacba)

simseulnyang commented:
코드 확인했습니다. 수고하셨습니다!

JaeHyuckSa commented:
추가적인 내용 (회의 정리 내용)
- [x] Post Model에 content_id 추가

Chestnut90 committed:
:recycle:Refact : Post 모델 리팩토링 #4

- content_id 필드 추가, social post에 대한 구분 식별자

Related to #4

Chestnut90 committed:
:recycle:Refact : 좋아요 api에 대한 content_id 적용

- post 모델의 content_id 필드가 추가됨에 따라서 likes/<content_id>/ 리소스 형태로 변경

Related to #11

Chestnut90 commented:
Post 모델에 대해 content_id를 uuid로 기본 설정하였습니다.
Serializer도 사용하는 field만 적용하도록 변경하였습니다.

saJaeHyukc commented:
> Post 모델에 대해 content_id를 uuid로 기본 설정하였습니다. Serializer도 사용하는 field만 적용하도록 변경하였습니다.

고생하셨습니다! 혹시 위에 code reply 적용해보실 수 있을까요?!

saJaeHyukc commented on config/urls.py:
```diff
@@ -24,6 +24,8 @@
     # Swagger
     path("swagger/docs/", schema_view.with_ui("swagger", cache_timeout=0), name="schema-swagger-ui"),
+    # Likes
+    path("likes/", include("likes.urls")),
```
저는 API 작성할 때 URL 규칙을 api/앱 이름/으로 설정했는데
` path("api/likes/", include("likes.urls")),` 로 변경해보는 것은 어떠실까요?

saJaeHyukc commented on likes/views.py:
```diff
@@ -1 +1,32 @@
+class LikesAPIView(APIView):
+    permission_classes = [IsAuthenticatedOrReadOnly]
+
+    def get_post(self, content_id):
+        try:
+            return Post.objects.get(content_id=content_id)
+        except Post.DoesNotExist:
+            raise NotFound("Post Not Found.")
+
+    def post(self, request, content_id):
+        post = self.get_post(content_id)
```
해당 코드를 not found를 처리하려는 코드로 확인됩니다! Django에서 지원하는 함수 중 get_object_or_404라는 함수가 있습니다 그것을 사용하면 코드를 한줄로 줄일 수 있을 것 같은데 한번 확인해보시면 어떨까요?? 아래 링크에서 한번 확인해보시면 좋을 것 같습니다!

https://developer-ankiwoong.tistory.com/956

Chestnut90 replied:
좋은 정보 감사합니다!

saJaeHyukc replied:
추가로 적용하셨다면 commit올리시고 재확인후 아래 resolve conversation을 누르도록 하겠습니다!

saJaeHyukc commented on config/urls.py:
```diff
@@ -24,6 +24,8 @@
     # Swagger
     path("swagger/docs/", schema_view.with_ui("swagger", cache_timeout=0), name="schema-swagger-ui"),
+    # Likes
+    path("likes/", include("likes.urls")),
```
저는 API 작성할 때 URL 규칙을 api/앱 이름/으로 설정했는데
` path("api/likes/", include("likes.urls")),` 로 변경해보는 것은 어떠실까요?

Chestnut90 replied:
네 좋은 방법입니다! 
버전에 대한 내용이 사전에 고려되지 않았었네요.
코멘트대로 domain/api/app 형식도 있고 domain/api/version/app 또는 api.domain/version/app 형태도 있는 것 같습니다.

https://www.freecodecamp.org/news/how-to-version-a-rest-api/
https://developers.naver.com/docs/common/openapiguide/apilist.md

saJaeHyukc replied:
오오 좋은 정보 감사합니다. 제 생각이지만 만약 실제 운영하는 서비스이면 버전링을 할 필요는 있을 것 같습니다! 원티드 과제 또는 버전업데이트는 추후 할까?라는 의문점이 있어 버전링까지는 적용하지 않았습니다. :)

Chestnut90 replied:
맞아요 저도 같은 생각이에요!

saJaeHyukc commented on likes/tests.py:
```diff
@@ -1 +1,37 @@
+
+
+class LikeAPITestCase(APITestCase):
+    client = APIClient(enforce_csrf_checks=True)
```
URL을 작성하실 때 django의 reverse를 사용하여 url에 앱네임을 지정하여 사용하시면 더 간단하게 확인할 수 있습니다! 
아래 작성한 코드를 보시고 한번 해보시면 어떨까요? 

https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/blob/develop/posts/tests/views/test_statics_list_view.py

saJaeHyukc commented on likes/urls.py:
```diff
@@ -0,0 +1,7 @@
+from django.urls import path
+
+from . import views
```
import 경로 규칙을 PEP 8에서 절대경로를 표시해주는 것을 권장하는 것으로 알고있습니다 !  한번 적용해보시면 좋을 것 같습니다!

<img width="855" alt="image" src="https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/assets/104830931/46bed38a-2a44-4a8d-bba7-25de7ddfe89d">

https://peps.python.org/pep-0008/

Chestnut90 replied:
좋은 정보 감사합니다! 
flake8, autoflake가 해당 내용을 검사하진 않나보네요😢 
나중에 PEP8에 대해서 더 읽어봐야겠습니다.

saJaeHyukc commented on likes/views.py:
```diff
@@ -1 +1,32 @@
+            return Post.objects.get(content_id=content_id)
+        except Post.DoesNotExist:
+            raise NotFound("Post Not Found.")
+
```
swagger 적용이 확인이 안되어 있어 아래 코드처럼 적용해보시면 어떨까요?

https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/blob/develop/posts/views.py

더 정확한 문서를 확인해보고 싶으시다면 아래 공식 문서를 참고해보시면 좋을 것 같습니다.

https://drf-yasg.readthedocs.io/en/stable/

Chestnut90 replied:
<img width="1399" alt="image" src="https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/assets/22581530/034fbe98-833b-43bb-afe4-2897f5dcecee">
제가 스웨거를 놓쳤군요! 
api 설명에 다른 추가적인 내용이나 일반적으로 사용하는 방법이 있는지 아시면 내용 공유하면 좋을 것 같아요!

saJaeHyukc commented on posts/models.py:
```diff
@@ -10,6 +12,7 @@ class PostType(models.TextChoices):
         INSTAGRAM = "instagram"
         THREADS = "threads"
 
+    content_id = models.UUIDField(default=uuid.uuid4, editable=False)
```
UUIDField로 하신 것은 정말 좋은 것 같습니다 기본 값으로 uuid4 함수를 통하여 자동 생성도 좋은 것 같습니다! :)

JaeHyuckSa reviewed (commented):
코드 리뷰

Chestnut90 committed:
:recycle:Refact : LikeApi의 post object getter 변경

- post 모델에 대한 getter는 django의 내장 함수로 대체 가능
- 가독성 증가

Related to #11

Chestnut90 committed:
:recycle:Refact : api url 포맷 일치 #11

- 해당 프로젝트의 url format을 api/{application}/ 로 사용하기로 정함.

Related to #11

Chestnut90 committed:
:recycle:Refact : API name 부여 및 testcase 리팩토링 #11

- 뷰에 대한 이름 지정
- API 테스트에서 url을 reverse를 통해 가져올 수 있도록
- 추후 뷰에 대한 변경이 파급이 적도록

Related to #11

Chestnut90 committed:
:recycle:Refact : 절대경로 import 변경 #11

- PEP8 권장에 따른 상대경로에서 절대경로로 변경함

Related to #11

Chestnut90 committed:
:recycle:Refact : likes api에 대한 swagger 적용 #11

- api 문서화를 위한 swagger 적용

Related to #11
